examples/input-element-default-checkedness/README.md

```diff
@@ -0,0 +1,38 @@
+# input-element-default-checkedness
+
+This example demonstrates an input element that maintains a **explicitly configured default checkedness** (i.e. does not change the `checked` HTML attribute) despite changes to `HTMLInputElement.checked`.
+
+`percy` does not support explicitly setting the default checkedness ([why?](#why-does-percy-override-default-checkedness)), so this also serves as an example of **appending independently-managed DOM elements to the DOM elements controlled by `percy-dom`'s virtual DOM**. Elements that are outside `percy-dom`'s virtual DOM model and will not be affected by it (unless, for example, the parent element is removed).
+
+---
+
+### When Should I Configure Default Checkedness Independently of Checkedness?
+
+This is irrelevant to server-side rendering, where there is no server-side DOM and only the HTML attributes are sent.
+
+This is not generally useful for client side rendering, as the default only matters if you want to configure what happens when checkedness isn't specified and `HTMLFormElement.reset()` is called or similarly, a reset element is pressed (which changes the element's checkedness to the default checkedness).
+- Most `percy` applications are expected to have a client-side application state, with GUI reflecting that app state. This is in contrast to UI-driven apps (which often scale poorly as complexity increases). In this case, the checkedness of an input is always specified according to app state.
+- Calling external functions that independently affect DOM elements such as `HTMLFormElement.reset()` may cause the DOM state may desync from the virtual DOM and app state, and hence is typically best avoided.
+
+---
+
+### Why Does `percy` Override Default Checkedness?
```

Comment:
> So, if you're using `percy-dom` strictly for server-side rendering, with no client-side logic, this would be a regression.

I'm having doubts that this was the right decision. Why would anyone ever use `percy-dom` for server-side rendering? The server shouldn't be manipulating a DOM server-side. That should be actively discouraged, right?

Therefore `percy-dom` has no reason to be overriding the default checkedness except to match `virtual-node`'s behavior, but why should it? They're doing two different things. One is encoding a virtual node tree as HTML, the other is creating, diffing, and patching DOM nodes based on two different virtual node trees.

At this point, this example seems like a whole lot of faffing because we're doing the wrong thing.

Proposal:
- Let's be explicit that `percy-dom` is not for server-side rendering.
- Let's have `percy-dom` explicitly do what makes sense for the context of the browser: `checked` manipulates the `HTMLInputElement.checked` property - it doesn't mess around with default checkedness as there's no reason to do so. People writing typical/idiomatic `percy` apps should never need to worry about it. And if for some reason they do, they shouldn't have to work around this strange behavior.

What we gain from keeping this behavior:
- `percy-dom` maintains closer HTML in the browser as it does to if you were to call `virtual_node.to_string()` (I don't see why this is useful, and can't think of a way this would cause issues. Can you?)

What is gained from tossing this behavior:
- Nobody needs to worry about what `percy-dom` does to default checkedness: nothing.
- `percy-dom` can avoid dealing with default checkedness entirely, as it's not useful for application-state-driven GUIs (which should always specify checkedness according to app state).
- `percy-dom` doesn't have an expectation of being used for something it shouldn't be used for.

---

I'd prefer to avoid "sunk cost fallacy" to kick in here and we just stick with what we have now. Unfortunately my understanding of the issue took a while to catch up (although there's a chance I'm still missing something of course).

Reply:
I don't remember where I wrote that quote, so I don't understand what you're referring to.

---

> The server shouldn't be manipulating a DOM server-side.

A DOM does require on the browser in any way. We should not couple these unrelated concepts.

Whether `percy` can manipulate a DOM should never depend on whether it is running in a browser.

I'll name some reasons:

- testing. It is an eventual goal that to be able to test the vast majority of your `percy` application outside of a browser. So, instead of using `web-sys`, we would use `generic-dom-traits` that used `web-sys` in a browser and
  something else outside of the browser.
  This would enable fast testing of things like "what happens when we click on this element but the onclick handler calls `event.prevent_default()`".
  For instance, we might use in-process-memory DOM implementation that was reliable enough that you trusted the results, and fast enough that one wouldn't hesitate to write as many test cases as they wanted
  - then a majority of an application's tests could be fast, and a minority that really needed to be in a browser could bear that overhead

- simulation / benchmarking / etc
  - if you want to run your application in an environment that has a DOM, you should be able to. We should not arbitrarily limit your application to only be able to run inside of probably-slow, and probably-cumbersome-to-use web browser

> it doesn't mess around with default checkedness as there's no reason to do so

It comes down to the framing. You're framing it as being about "default checkedness".
In that context, yes we currently have no reason to believe that we need `percy-dom` to do anything to the "default checkedness".

However, if this is framed around "setting the checked attribute" and "the user thinks that they're setting the checked attribute", then that changes things.

So, the question becomes, would a user ever want to set the "checked" attribute?

If `<input foo="bar" />` sets the `foo` attribute, a user might expect `<input checked=true />` to set the "checked" attribute.

If we pointed the user at an application and asked "what do you think happens when we render `<input checked=true />` they should probably say "The checked attribute gets set".

In practice this would only impact the user if they were doing `input.getAttribute('checked')` and got back `null` instead of `"true"`.

It's entirely possible that roughly 0 people will ever do this.

And, if they do, they can open an issue and we can do further design work.

So, I'm okay with, for now, having `percy-dom` only control the `checked` property, and not the `checked` attribute.

## Proposal

> Let's be explicit that percy-dom is not for server-side rendering.

Don't do this. Instead, we can say that if a user truly wants to set the "checked" attribute they can use `.setAttribute` on the node themselves https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute .

> checked manipulates the HTMLInputElement.checked property - it doesn't mess around with ~default checkedness~ *checked attribute* (words swapped by me)

Sounds good.

Reply:
> Sounds good.

Neat.

I'll delete the example as it's no longer illustrative of doing something reasonable. If users want to set the default checkedness of an element, they should just get the DOM element and add/remove the attribute. It can be grabbed from the commit history if its useful later?

It's also worth mentioning that the situation with `value` is _exactly_ the same as with `checked`, and having different behavior for them would be strange, at least long-term. If nothing explodes by releasing this change out into the wild, I think changing how `percy-dom` treats `value` should be changed to match `checked`. For now, the book entries reflect the difference.

> Let's be explicit that percy-dom is not for server-side rendering.

I will leave it out. It was only in service of motivating the `checked` VDOM attribute behavior.

> However, if this is framed around "setting the checked attribute" and "the user thinks that they're setting the checked attribute", then that changes things.

It's unfortunate.
1. The core issue is the HTML standard. The `checked` HTML attribute should be called `defaultChecked` or something to distinguish it.
2. Secondly, using a macro called `html!` with a HTML-like format to manipulate a DOM. It's misleading because an HTML document and a DOM have aliased but distinct concepts, such as the `checked` attribute and the `checked` property. ...It's a small enough issue that the choice might be worth the benefit of the intuitive markup.

Unless we remove the `checked` attribute from `percy` (and use something like `is_checked` and `default_checked` just to make it harder to assume what they mean) I think we're going to need to bite the bullet that devs will sometimes assume that `percy-dom` does the wrong thing - with respect to `percy`'s intended utility.

But at least `percy-dom` will be doing the sensible thing (as far as we can tell, at least), so it's "just" a learning speed bump. At least `percy` isn't hard to use once the dev (hopefully) reads the book section to figure out what's going on.

---

Some leftover thoughts:

> A DOM does require on the browser in any way. We should not couple these unrelated concepts.
> 
> Whether percy can manipulate a DOM should never depend on whether it is running in a browser.

I agree. We're talking past each other though. I don't think this is relevant to what I'm proposing.

Testing, simulation, benchmarking, etc. in environments without a browser seem neat. I just don't see what this has to do with server-side rendering though.

Server-side rendering is: generating a static HTML document and sending it off to the client. Manipulating a DOM is not useful in this process. Manipulating a virtual DOM is useful. But `percy` already has a distinction between `virtual-node`+`html-macro` and `percy-dom` such that manipulating a virtual DOM server-side to generate HTML has no dependency on the idea of actually manipulating a real or simulated or mocked DOM.

Reply:
> I'll delete the example

Unless it would be hard, let's just change it to be a `embed-non-percy-node` example where we show you how to create your own node that `percy` doesn't touch.

Can just delete all of the checkedness stuff and instead just make the example create a `<div hello="world" />` or something using `web-sys` and insert that into
a percy contain `.on_create_element`.

> If nothing explodes by releasing this change out into the wild, I think changing how percy-dom treats value should be changed to match checked

Sounds good.

> Secondly, using a macro called html! with a HTML-like format to manipulate a DOM

`html!` macro is describing the DOM that you want.

Then `percy-dom` manipulates it.

The macro isn't manipulating the DOM, just declaring one.

So, it's designed to stay close to how a user would expect to be able to declare a DOM tree. Since HTML is likely the most common way to way to define a DOM tree, it's designed to feel like HTML.

This lets users avoid learning percy and instead just learn common specs/concepts. Transferable knowledge.

> Unless we remove the checked attribute from percy (and use something like is_checked and default_checked just to make it harder to assume what they mean)
> I think we're going to need to bite the bullet that devs will sometimes assume that percy-dom does the wrong thing - with respect to percy's intended utility.

_rough idea_

I could see a potential future (would require design work) where the `html!` macro gave a compile time error if you used `checked`.

It would tell you to use `is_checked` or how to set the attribute if you really wanted to do that.

That wouldn't help users that use `VirtualNode` directly, however.

> But at least percy-dom will be doing the sensible thing (as far as we can tell, at least), so it's "just" a learning speed bump

Yeah, especially if the macro compile time error points you to the docs.

> I agree. We're talking past each other though

Yeah I only realized that after I continued reading your comment

> Server-side rendering is: generating a static HTML document and sending it off to the client. Manipulating a DOM is not useful in this process.

Here's what I had in mind:

- user starts some sort of simulation application where they are running their percy application in an emulated browser

- users triggers a bunch of clicks / input events / etc

- user now wants to take the DOM tree from that emulated browser and turn it into a string

If you're really on default checkedness that DOM tree will be different from your virtual tree. Your virtual tree doesn't know whether the element is checked. Only the emulated DOM has info about whether the input is checked.

I haven't thought enough about this to know if it relates to our convo, so I'm more so communicating that I'm not certain that we should operate as if server-side rendering of a "real DOM" will never be useful.